src/simulators/tensor_network/MPS.cpp

```diff
@@ -0,0 +1,398 @@
+		apply_swap(index_A-1,index_B);
+		return;
+	}
+
+	bool swapped = false;
+	if(index_A >  index_B)
+	{
+		myswap<uint_t>(index_A, index_B);
+		swapped = true;
+	}
+
+	MPS_Tensor A = q_reg_[index_A], B = q_reg_[index_B];
+	rvector_t left_lambda, right_lambda;
+	//There is no lambda in the edges of the MPS
+	left_lambda  = (index_A != 0) 	    ? lambda_reg_[index_A-1] : rvector_t {1.0};
+	right_lambda = (index_B != num_qubits_-1) ? lambda_reg_[index_B  ] : rvector_t {1.0};
+
+	q_reg_[index_A].mul_Gamma_by_left_Lambda(left_lambda);
+	q_reg_[index_B].mul_Gamma_by_right_Lambda(right_lambda);
+	MPS_Tensor temp = MPS_Tensor::contract(q_reg_[index_A], lambda_reg_[index_A], q_reg_[index_B]);
+
+	if(DEBUG) temp.print();
+	switch (gate_type) {
+	case cx:
+	  temp.apply_cnot(swapped);
+	  break;
+	case cz:
+	  temp.apply_cz();
+	  break;
+	case cu:
+	{
+	  cmatrix_t Zeros = AER::Utils::Matrix::I-AER::Utils::Matrix::I;
+	  cmatrix_t temp1 = AER::Utils::concatenate(AER::Utils::Matrix::I, Zeros , 1),
+			  	temp2 = AER::Utils::concatenate(Zeros, mat, 1);
+	  cmatrix_t cu = AER::Utils::concatenate(temp1, temp2 ,0) ;
+	  temp.apply_matrix(cu);
+	  break;
+	}
+	case su4:
+	  temp.apply_matrix(mat);
+	  break;
+
+    default:
+	  throw std::invalid_argument("illegal gate for apply_2_qubit_gate"); 
+	}
+	if(DEBUG) temp.print();
+	MPS_Tensor left_gamma,right_gamma;
+	rvector_t lambda;
+	MPS_Tensor::Decompose(temp, left_gamma, lambda, right_gamma);
+	left_gamma.div_Gamma_by_left_Lambda(left_lambda);
+	right_gamma.div_Gamma_by_right_Lambda(right_lambda);
+	q_reg_[index_A] = left_gamma;
+	lambda_reg_[index_A] = lambda;
+	q_reg_[index_B] = right_gamma;
+}
+
+void MPS::change_position(uint_t src, uint_t dst)
+{
+	if(src == dst)
+		return;
+	else if(src < dst)
+		for(uint_t i = src; i < dst; i++)
+			apply_swap(i,i+1);
+	else
+		for(uint_t i = src; i > dst; i--)
+			apply_swap(i,i-1);
+}
+
+cmatrix_t MPS::Density_matrix(const reg_t &qubits) const
+{
+  // ***** Assuming ascending sorted qubits register *****
+  vector<uint_t> internalIndexes;
+  for (uint_t index : qubits)
+    internalIndexes.push_back((uint_t)index);
+  //  std::sort(internalIndexes.begin(), internalIndexes.end()); -- Assuming sorted
+
+  MPS temp_TN;
+  temp_TN.initialize(*this);
+  vector<uint_t> new_indexes = calc_new_indexes(internalIndexes);
+  uint_t avg = new_indexes[new_indexes.size()/2];
+  vector<uint_t>::iterator it = lower_bound(internalIndexes.begin(), internalIndexes.end(), avg);
+  int mid = std::distance(internalIndexes.begin(), it);
+  for(uint_t i = mid; i < internalIndexes.size(); i++)
+  {
+    temp_TN.change_position(internalIndexes[i],new_indexes[i]);
+  }
+  for(int i = mid-1; i >= 0; i--)
+  {
+    temp_TN.change_position(internalIndexes[i],new_indexes[i]);
+  }
+  MPS_Tensor psi = temp_TN.state_vec(new_indexes.front(), new_indexes.back());
+  uint_t size = psi.get_dim();
+  cmatrix_t rho(size,size);
+  for(uint_t i = 0; i < size; i++) {
```

Comment:
This seems to be a good place for parallelizing.
Can you add: 
```c++
    #ifdef _WIN32
      #pragma omp for
    #else
      #pragma omp for collapse(2)
    #endif
    for(int_t i = 0; i < size; i++) {
 ```
Remember to change the `uint_t` from the `for(uint_t ...)` to `int_t` (due to the lack of proper support for OMP on Windows).

Reply:
Done.